### PR TITLE
Allow a consumer to switch between JSON and binary encoding for awareness

### PR DESCRIPTION
**Pull request by Myrannas**
This PR is intended to add support for binary awareness encoding as discussed in #26. I'd appreciate any feedback you have!

I have tested it out locally for my use case - and it appears to be working correctly in both modes.

### CODE REVIEW

File: awareness.js [hunk 1]
```diff
@@ -156,6 +156,54 @@ export class Awareness extends Observable {
...
+ *
+ * This is the default strategy used in this package.
+ */
+export class DefaultAwarenessStateEncoder {
+  /**
+   * Encode an awareness state entry
+   * @param {encoding.Encoder} encoder
+   * @param {any} update
+   */
+  encodeState (encoder, update) {
```
Comment (dmonad) [inline awareness.js:170]: These should be `static` methods. Then there is no need to create an instance of `*AwarenessStateEncoder`

File: awareness.js [hunk 2]
```diff
@@ -189,9 +237,10 @@ export const removeAwarenessStates = (awareness, clients, origin) => {
 /**
  * @param {Awareness} awareness
  * @param {Array<number>} clients
+ * @param {DefaultAwarenessStateEncoder|BinaryAwarenessStateEncoder} stateEncoder The encoder to use for encoding and decoding each state entry
  * @return {Uint8Array}
  */
-export const encodeAwarenessUpdate = (awareness, clients, states = awareness.states) => {
+export const encodeAwarenessUpdate = (awareness, clients, states = awareness.states, stateEncoder = new DefaultAwarenessStateEncoder()) => {
```
Comment (dmonad) [inline awareness.js:243]: It should not be necessary to create a new instance of anything here. `*AwarenessEncoder` should only have static methods.

File: awareness.test.js [hunk 3]
```diff
@@ -51,3 +52,7 @@ export const testAwareness = tc => {
   t.compare(aw1.getStates().get(0), undefined)
   t.compare(lastChangeLocal, lastChange)
 }
+
+export const testAwareness = testAwarenessWithEncoding()
```
Comment (dmonad) [inline awareness.test.js:56]: This test is the same as the third test and can be removed.

**Review by dmonad — COMMENTED** [review]
Sorry, I merged to quickly. I just reverted the changes. Could you please make the suggested changes before I make a release? You probably have to create a new PR.

File: awareness.js (hunk 1 above)
Comment (Myrannas) [reply awareness.js:170]: Sure, can do.

File: awareness.test.js (hunk 3 above)
Comment (Myrannas) [reply awareness.test.js:56]: I'll augment this test a bit, I'd like to add a test to ensure the default behaviour is using the default encoder.

**Pull request merged**